src/nvim/api/vimscript.c

```diff
@@ -745,3 +746,294 @@ Dictionary nvim_parse_expression(String expr, String flags, Boolean highlight, E
+        }
+        kv_push(args, DICTIONARY_OBJ(arg));
+      }
+    }
+    PUT(dict, "args", ARRAY_OBJ(args));
+    PUT(dict, "varargs", BOOLEAN_OBJ(fp->uf_varargs));
+    // Attributes
+    PUT(dict, "abort", BOOLEAN_OBJ(fp->uf_flags & FC_ABORT));
+    PUT(dict, "range", BOOLEAN_OBJ(fp->uf_flags & FC_RANGE));
+    PUT(dict, "dict", BOOLEAN_OBJ(fp->uf_flags & FC_DICT));
+    PUT(dict, "closure", BOOLEAN_OBJ(fp->uf_flags & FC_CLOSURE));
+    // Script
+    PUT(dict, "sid", INTEGER_OBJ(fp->uf_script_ctx.sc_sid));
+    PUT(dict, "lnum", INTEGER_OBJ(fp->uf_script_ctx.sc_lnum));
+  }
+
+  // Source lines
+  if (lines) {
+    Array array = ARRAY_DICT_INIT;
+    if (fp->uf_lines.ga_len > 0) {
+      kv_resize(array, (size_t)fp->uf_lines.ga_len);
+      for (int j = 0; j < fp->uf_lines.ga_len; j++) {
+        kv_push(array, CSTR_TO_OBJ((const char *)FUNCLINE(fp, j)));
+      }
+    }
+    PUT(dict, "lines", ARRAY_OBJ(array));
+  }
+
+  return dict;
+}
+
+/// Takes ownership of name.
+static Dictionary builtin_function_dict(const EvalFuncDef *fn, String name, bool details)
+{
+  Dictionary dict = ARRAY_DICT_INIT;
+  kv_resize(dict, details ? 7 : 2);
+
+  PUT(dict, "name", STRING_OBJ(name));
+  PUT(dict, "type", CSTR_TO_OBJ("builtin"));
+
+  if (details) {
+    PUT(dict, "min_argc", INTEGER_OBJ(fn->min_argc));
+    PUT(dict, "max_argc", INTEGER_OBJ(fn->max_argc));
+    PUT(dict, "base_arg", INTEGER_OBJ(fn->base_arg));
+    PUT(dict, "fast", BOOLEAN_OBJ(fn->fast));
+
+    // Argument names
+    if (fn->argnames != NULL) {
+      Array overloads = ARRAY_DICT_INIT;
+      Array names = ARRAY_DICT_INIT;
+      char buf[64] = { 0 };
+      size_t pos = 0;
+
+      // Argument are separated with unit separator (ascii \x1F),
+      // overloads with record separator (ascii \x1E).
+      for (const char *p = fn->argnames;; p++) {
+        if (*p == NUL || *p == '\x1F' || *p == '\x1E') {
+          assert(pos < 64);
+          buf[pos++] = NUL;
+          String str = {
+            .data = xmemdupz(buf, pos),
+            .size = pos - 1,
+          };
+          pos = 0;
+
+          if (*p == '\x1F') {  // unit separator: argument separator
+            kv_push(names, STRING_OBJ(str));
+          } else if (*p == '\x1E') {  // record separator: overload separator
+            kv_push(names, STRING_OBJ(str));
+            kv_push(overloads, ARRAY_OBJ(names));
+            names = (Array)ARRAY_DICT_INIT;
+          } else {  // NUL
+            kv_push(names, STRING_OBJ(str));
+            kv_push(overloads, ARRAY_OBJ(names));
+            break;
+          }
+        } else {
+          buf[pos++] = *p;
+        }
+      }
+
+      PUT(dict, "argnames", ARRAY_OBJ(overloads));
+    }
+  }
+
+  return dict;
+}
+
+/// Lists Vimscript functions.
+///
+/// @param query    When string gets information about the function under this name.
+///                 When dictionary lists functions. The following keys are accepted:
```

Comment:
i don't understand the dictionary case. Need some rewording here.

Reply:
Dictionary is `{ builtin = true/false, user = true/false }`, and it returns a list of functions of requested types (builtin/user).

I'm also not 100% sure if switching the behavior between "get one function" and "list all functions" based on the argument type is a right thing to do here. I wonder if it wouldn't be better to take a single dictionary argument and make it consistent with `nvim_get_commands`. Both API functions could act like this:
- `nvim_get_*({ name = 'foo' })` -> returns a single function/command "foo"
- `nvim_get_*({ user = true })` -> lists all user functions/commands
- `nvim_get_*({ builtin = true })` -> lists all builtin functions/commands
- `nvim_get_*({ user = true, builtin = true })` -> lists all functions/commands, both builtin and user
- `nvim_get_*({ name = 'foo', user = true })` -> error, can't use both at the same time